lib/src/appengine_internal.dart

```diff
@@ -410,3 +409,20 @@ class StderrLoggerFactory implements LoggerFactory {
     return stderr_logging_impl.StderrBackgroundLoggingImpl();
   }
 }
+
+class _DatastoreDBFactoryImpl implements DatastoreDBFactory {
+  _DatastoreDBFactoryImpl(this.projectId, this.endpoint, this.needAuthorization, this.authenticator);
+
+  final String projectId;
+  final String endpoint;
+  final bool needAuthorization;
+  final grpc.HttpBasedAuthenticator authenticator;
+
+  @override
+  db.DatastoreDB newDatastoreDB() {
+    final grpcClient = _getGrpcClientChannel(endpoint, needAuthorization);
```

Comment:
This might be a bit problematic..

From what I can see this gets called from `_contextRegistry._getServices` which is called from `_contextRegistry.add` which is called before we do `ss.fork(...`, so each of these clients are created in the global service scope, and not in a per-request service scope. Hence, the TCP connection opened by the GRPC client will not be closed until it's garbage collected. That's likely to causing significant resource constraints on servers processing many requests.

However, even if we were to close the TCP connection after the request is done, this would still be undesirable. Because:
  (A) It would increase the latency for every request using datastore as a new TCP connection needs to be established.
  (B) More critically with around 28k ephemeral ports on Linux, and old ports remaining in TIME_WAIT for 120 seconds after use, **the server would run out of ephemeral ports** for creating new outgoing connections after a 2-3 minutes with a load of ~230 req/s. This would cause increasing wait time for new TCP connections and gradually degrade the server until establishing new TCP connections takes around 120s. Usually, the server will then crash under the memory load of incoming requests held pending.

Note, that (B) is a problem here because we're connecting to the same server (or at-least we very likely are connecting to the same server or set of servers).

@tvolkert, am I missing something here?

Reply:
Nope, that analysis sounds right - I'll revert and publish as 0.10.0.  Then I'd love to strategize on how to solve the underlying problem that this PR attempted to solve (concurrent requests stomping on each others' grpc:Http2ClientConnection)